client/widgets/markets/CArtifactsBuying.cpp

```diff
@@ -0,0 +1,124 @@
+
+	// Player's resources
+	assert(bidTradePanel);
+	bidTradePanel->moveTo(pos.topLeft() + Point(39, 184));
+	bidTradePanel->showcaseSlot->image->moveTo(pos.topLeft() + Point(141, 454));
+
+	// Artifacts panel
+	offerTradePanel = std::make_shared<ArtifactsPanel>([this](const std::shared_ptr<CTradeableItem> & newSlot)
+		{
+			CArtifactsBuying::onSlotClickPressed(newSlot, offerTradePanel);
+		}, [this]()
+		{
+			CMarketBase::updateSubtitlesForBid(EMarketMode::RESOURCE_ARTIFACT, bidTradePanel->getSelectedItemId());
+		}, market->availableItemsIds(EMarketMode::RESOURCE_ARTIFACT));
+	offerTradePanel->deleteSlotsCheck = [this](const std::shared_ptr<CTradeableItem> & slot)
+	{
+		return vstd::contains(this->market->availableItemsIds(EMarketMode::RESOURCE_ARTIFACT), ArtifactID(slot->id)) ? false : true;
+	};
+	offerTradePanel->moveTo(pos.topLeft() + Point(328, 181));
+
+	CMarketBase::update();
+	CArtifactsBuying::deselect();
+}
+
+void CArtifactsBuying::deselect()
+{
+	CMarketBase::deselect();
+	CMarketTraderText::deselect();
+}
+
+void CArtifactsBuying::makeDeal()
+{
+	if(ArtifactID(offerTradePanel->getSelectedItemId()).toArtifact()->canBePutAt(hero))
+	{
+		LOCPLINT->cb->trade(market, EMarketMode::RESOURCE_ARTIFACT, GameResID(bidTradePanel->getSelectedItemId()),
+			ArtifactID(offerTradePanel->getSelectedItemId()), offerQty, hero);
+		CMarketTraderText::makeDeal();
+		deselect();
+	}
+	else
+	{
+		LOCPLINT->showInfoDialog(CGI->generaltexth->translate("core.genrltxt.326"));
+	}
+}
+
+CMarketBase::MarketShowcasesParams CArtifactsBuying::getShowcasesParams() const
+{
+	if(bidTradePanel->isHighlighted() && offerTradePanel->isHighlighted())
+		return std::make_tuple(
+			ShowcaseParams {std::to_string(deal->isBlocked() ? 0 : bidQty), bidTradePanel->getSelectedItemId()},
+			ShowcaseParams {std::to_string(deal->isBlocked() ? 0 : offerQty), CGI->artifacts()->getByIndex(offerTradePanel->getSelectedItemId())->getIconIndex()});
+	else
+		return std::make_tuple(std::nullopt, std::nullopt);
+}
+
+void CArtifactsBuying::highlightingChanged()
+{
+	if(bidTradePanel->isHighlighted() && offerTradePanel->isHighlighted())
+	{
+		market->getOffer(bidTradePanel->getSelectedItemId(), offerTradePanel->getSelectedItemId(), bidQty, offerQty, EMarketMode::RESOURCE_ARTIFACT);
+		deal->block(LOCPLINT->cb->getResourceAmount(GameResID(bidTradePanel->getSelectedItemId())) >= bidQty ? false : true);
+	}
+	CMarketBase::highlightingChanged();
+	CMarketTraderText::highlightingChanged();
+}
+
+std::string CArtifactsBuying::getTraderText()
+{
+	if(bidTradePanel->isHighlighted() && offerTradePanel->isHighlighted())
+	{
+		return boost::str(boost::format(
+			CGI->generaltexth->allTexts[267]) %
+			CGI->artifacts()->getByIndex(offerTradePanel->getSelectedItemId())->getNameTranslated() %
+			bidQty %
+			(bidQty == 1 ? CGI->generaltexth->allTexts[161] : CGI->generaltexth->allTexts[160]) %
+			CGI->generaltexth->restypes[bidTradePanel->getSelectedItemId()]);
```

Comment:
While I am not aware of issues with this particular string, but our MetaString is probably more preferrable to boost::format, at least for new code.

Boost format throws exceptions when format string contains less placeholders ('%s') than provided parameters. But some H3 localization sometimes remove placeholders from translated strings. 

Our MetaString avoids such issue by silently ignoring excessive replacements and might have more readable syntax since it is more integrated with our classes, e.g:
```cpp
MetaString message = MetaString::fromTextID("core.genrltxt.267");
message.replaceName(ArtifactID(offerTradePanel->getSelectedItemId()));
message.replaceNumber(bidQty);
message.replaceTextID( bidQty == 1 ? "core.genrltxt.161" : "core.genrltxt.160")
message.replaceName(GameResID(bidTradePanel->getSelectedItemId())));
return message.toString();
```